### PR TITLE
feat: handle multiple grpc deletes

### LINKED ISSUE #1708: Optimize distributed deletions from gRPC
### What type of enhancement is this?

Refactor

### What does the enhancement do?

GRPC delete requests are just like insert requests, carrying the column data for deletions. They both had very similar implementations: splitting the to be inserted/deleted data by partition rules, not parallel or batched, and done in a per table fashion etc.

Now the distributed insertions are optimized in a batch in https://github.com/GreptimeTeam/greptimedb/pull/1687 , it's time we do that for distributed deletions as well.

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by NiwakaDev**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This PR allows `greptimedb` to handle multiple grpc deletes.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)

close #1708

### CODE REVIEW

**Comment by NiwakaDev** [comment]
It seems like `CI / Sqlness Test (windows-latest)` fails consistently, like so:
```
Going to build the DB...
thread 'main' panicked at tests\runner\src\env.rs:319:13:
Failed to build GreptimeDB (`cargo build` fails)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: process didn't exit successfully: `target\debug\sqlness-runner.exe` (exit code: 101)
Error: Process completed with exit code 1.
```

**Comment by MichaelScofield** [comment]
@NiwakaDev I've made the cargo build in sqlness prints the output if it fails in  https://github.com/GreptimeTeam/greptimedb/pull/2152 . After that PR is merged, plz rebase your code and see how the ci fails

**Comment by MichaelScofield** [comment]
@NiwakaDev disk is full. plz wait for https://github.com/GreptimeTeam/greptimedb/pull/2158

File: src/partition/src/manager.rs [hunk 1]
```diff
@@ -246,6 +246,17 @@ impl PartitionRuleManager {
...
         splitter.split_insert(req, schema)
     }
+
+    pub async fn split_delete_request(
+        &self,
+        table: TableId,
+        req: DeleteRequest,
+        primary_key_column_names: Vec<&String>,
+    ) -> Result<DeleteRequestSplit> {
+        let partition_rule = self.find_table_partition_rule(table).await.unwrap();
```
Comment (MichaelScofield) [inline src/partition/src/manager.rs:256]: We can simply propagate the error:

```suggestion
        let partition_rule = self.find_table_partition_rule(table).await?;
```

File: src/frontend/src/instance/distributed.rs [hunk 2]
```diff
@@ -626,28 +627,16 @@ impl DistInstance {
...
-            .context(ToTableDeleteRequestSnafu)?;
-
-        let affected_rows = table.delete(request).await.context(TableSnafu)?;
-        Ok(Output::AffectedRows(affected_rows))
+        let deleter = DistDeleter::new(
+            ctx.current_catalog().to_string(),
+            ctx.current_schema().to_string(),
+            self.catalog_manager(),
+        );
+        let affected_rows = deleter.grpc_delete(request).await?;
```
Comment (killme2008) [inline src/frontend/src/instance/distributed.rs:638]: Make `grpc_delete` return the `usize` type directly to avoid casting.

**Review by killme2008 — APPROVED** [review]
Almost LGTM

File: src/frontend/src/instance/distributed.rs (hunk 2 above)
Comment (NiwakaDev) [reply src/frontend/src/instance/distributed.rs:638]: Done: [222c098](https://github.com/GreptimeTeam/greptimedb/pull/2150/commits/222c0985a930c41a00050f1a7b05938bef7ab9cd)

Please let me know if there are any errors or omissions.

File: src/partition/src/manager.rs (hunk 1 above)
Comment (NiwakaDev) [reply src/partition/src/manager.rs]: Done: [5f27dd8](https://github.com/GreptimeTeam/greptimedb/pull/2150/commits/5f27dd85057ba9868eb42cd95afd63b7d2d362c2)

Please let me know if there are any errors or omissinos.

File: src/frontend/src/error.rs [hunk 3]
```diff
@@ -409,6 +415,12 @@ pub enum Error {
         source: table::error::Error,
     },
 
+    #[snafu(display("Missing time index column: {}", source))]
+    MissingTimeIndexColumn {
+        #[snafu(backtrace)]
```
[the remark this answers is not in the document]
Comment (NiwakaDev) [reply src/frontend/src/error.rs]: Done: [2c90952](https://github.com/GreptimeTeam/greptimedb/pull/2150/commits/2c909523cd0c17ed2633b6db0bb87f3dc7326ab6)

**Review by MichaelScofield — COMMENTED** [review]
LGTM. Please resolve the conflicts so I can merge it.

**Comment by NiwakaDev** [comment]
@MichaelScofield 

>Please resolve the conflicts so I can merge it.

I resolved them.

**Review by MichaelScofield — APPROVED** [review]

**Comment by MichaelScofield** [comment]
@NiwakaDev thx!

**Pull request merged**